core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java

```diff
@@ -134,21 +135,150 @@ public void testCloudSqlInstanceForcesRefresh() throws InterruptedException {
+    instance.getSslData();
+    assertThat(refreshCount.get()).isEqualTo(1);
+
+    // Wait for the second refresh operation to complete
+    cond.proceedWhen(() -> refreshCount.get() > 1);
+
+    // getSslData again, and assert the refresh operation completed.
     instance.getSslData();
-    // refresh count hasn't changed because we re-use the existing connection info
+    assertThat(refreshCount.get()).isEqualTo(2);
+  }
+
+  @Test
+  public void testCloudSqlRefreshesExpiredData() throws Exception {
+    InstanceData aboutToExpireData =
+        new InstanceData(
+            null, new SslData(null, null, null), Instant.now().plus(2, ChronoUnit.SECONDS));
+
+    InstanceData data =
+        new InstanceData(
+            null, new SslData(null, null, null), Instant.now().plus(1, ChronoUnit.HOURS));
+
+    AtomicInteger refreshCount = new AtomicInteger();
+    final PauseCondition refresh1 = new PauseCondition();
+
+    CloudSqlInstance instance =
+        new CloudSqlInstance(
+            "project:region:instance",
+            (instanceName, accessTokenSupplier, authType, executor, keyPair) -> {
+              int c = refreshCount.get();
+              InstanceData refreshResult = data;
+              // Execution 1 and 2 should pause
+              switch (c) {
+                case 0:
+                  refreshResult = aboutToExpireData;
+                  break;
+                case 1:
+                  refresh1.pause();
+                  break;
+              }
+              refreshCount.incrementAndGet();
+              return refreshResult;
+            },
+            AuthType.PASSWORD,
+            stubCredentialFactory,
+            executorService,
+            keyPairFuture,
+            TEST_RATE_LIMITER);
+
+    // Get the first data that is about to expire
+    SslData d = instance.getSslData();
+    assertThat(refreshCount.get()).isEqualTo(1);
+    assertThat(d).isSameInstanceAs(aboutToExpireData.getSslData());
+
+    // Wait for the instance to expire
+    while (Instant.now().isBefore(aboutToExpireData.getExpiration())) {
+      Thread.sleep(10);
+    }
+
+    // Now that the InstanceData has expired, getInstanceData will return the same, expired
+    // token until a new one is retrieved.
     assertThat(refreshCount.get()).isEqualTo(1);
+    assertThat(instance.getSslData()).isSameInstanceAs(aboutToExpireData.getSslData());
+
+    // Allow the second refresh operation to complete
+    refresh1.proceedWhen(() -> refreshCount.get() > 1);
+    Thread.sleep(10);
+
+    // getSslData again, and assert the refresh operation completed.
+    SslData d2 = instance.getSslData();
+    assertThat(d2).isSameInstanceAs(data.getSslData());
+  }
+
+  @Test
+  public void testThatForceRefreshBalksWhenAForceRefreshIsInProgress() throws Exception {
+    InstanceData aboutToExpireData =
+        new InstanceData(
+            null, new SslData(null, null, null), Instant.now().plus(1, ChronoUnit.HOURS));
```

Comment:
it should be `ChronoUnit.SECONDS`

Reply:
In this testcase, we want the initial data to have a long expiration, so that it does not immediately trigger a refresh. Then, we want to call forceRefresh() and see that exactly 1 refresh operation occurs. 

I have renamed the variable `initialData` instead of `aboutToExpireData` for clarity.